### ISSUE TITLE
[BUG] DeserializeEventGridEvents is not working when using the ServiceBusQueue flow

### ISSUE THREAD

**Issue by jsgoupil** [Azure/azure-sdk-for-net#10137]
**Describe the bug**
I have not tried all consumer, but when trying it with the Azure Service Bus Queue, the data does not come as an array as mentioned in the doc page here https://docs.microsoft.com/en-us/dotnet/api/overview/azure/eventgrid?view=azure-dotnet

Consequently, calling `DeserializeEventGridEvents` expects to receive an array, but I receive only 1 object.

**Expected behavior**
Explain how this should work better via the documentation, or provide a better method, or fix the what is being passed in to the queue?

**Actual behavior (include Exception or Stack Trace)**
![image](https://user-images.githubusercontent.com/808649/75209549-5f2db100-5733-11ea-8720-e1ff6cf5f6c8.png)

**Environment:**
 - Microsoft.Azure.EventGrid: 3.2.0
 - Microsoft.Azure.ServiceBus: 3.1.1

**Comment by lilyjma** [comment]
@jfggdl, do we have any updates on this issue?

**Comment by ahamad-MS** [comment]
Hello @jsgoupil .. thanks for reaching out and reporting this issue. 

I am not sure if you are reporting the issue against the EventGridSubscriber.DeserializeEventGridEvents() Method (https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.eventgrid.eventgridsubscriber.deserializeeventgridevents?view=azure-dotnet ) or you are hitting this issue when you are trying to deserialize it yourself. From the code snippet above, it seems it is the latter case.

Reported
var evt = JsonConvert.DeserializeObject<EventGridEvent>(requestContent)

while I would expect something like

EventGridEvent[] eventGridEvents = JsonConvert.DeserializeObject<EventGridEvent[]>(requestContent)

In other words, you are missing [] in <EventGridEvent> and instead you would use <EventGridEvent[]>

Can you please try it out and if this does not answer you question, can you please provide additional information.

thanks

**Comment by jsgoupil** [comment]
No, it's the first case. 
The data coming in is not an array

**Comment by ahamad-MS** [comment]
Deliverying to Service bus queue does not support batching of events since service bus queue itself does not support that. You can try event hub for that matter.

Are you getting specific error when trying to EventGridSubscriber.DeserializeEventGridEvents() ?

**Comment by jsgoupil** [comment]
I understand that you might only support delivering only 1 event to the service bus, but then your API EventGridSubscriber does not work properly. It expects an array.

Here is the callstack.

```
Cannot deserialize the current JSON object (e.g. {"name":"value"}) into type 'Microsoft.Azure.EventGrid.Models.EventGridEvent[]' because the type requires a JSON array (e.g. [1,2,3]) to deserialize correctly.
To fix this error either change the JSON to a JSON array (e.g. [1,2,3]) or change the deserialized type so that it is a normal .NET type (e.g. not a primitive type like integer, not a collection type like an array or List<T>) that can be deserialized from a JSON object. JsonObjectAttribute can also be added to the type to force it to deserialize from a JSON object.
Path 'id', line 1, position 6.

   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
   at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)
   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value, JsonSerializerSettings settings)
   at Microsoft.Azure.EventGrid.EventGridSubscriber.DeserializeEventGridEvents(String requestContent, JsonSerializer jsonSerializer)
   at Microsoft.Azure.EventGrid.EventGridSubscriber.DeserializeEventGridEvents(String requestContent)
   at MyCode
```

**Comment by ahamad-MS** [comment]
Thanks for the additional info. it is helpful. We are adding this to our backlog and will triage appropriately. thanks.

**Comment by ahamad-MS** [comment]
We added this to our backlog and will address this appropriately.

**Comment by SeanFeldman** [comment]
> We added this to our backlog and will address this appropriately.

@ahamad-MS please re-open the issue. External folks don't have any visibility into internal backlogs. Unless there's an associated public issue, which you should link to, this should remail opened.

When the problem is addressed/fixed then it will be appropriate to close this (public) issue. Until then, anyone running into this problem should see that it's in-progress/backlogged. Thank you.

**Comment by ahamad-MS** [comment]
Update: This issue is being handled as part of the Azure Data Plane SDK integration and it is still in the design phase. We will be handling this issue/bug but no ETA for its availability yet.

**Comment by JoshLove-msft** [comment]
Just noting that this is not an issue in the new Event Grid SDK that is currently in beta - https://www.nuget.org/packages/Azure.Messaging.EventGrid/4.0.0-beta.2

**Comment by SeanFeldman** [comment]
> Just noting that this is not an issue in the new Event Grid SDK that is currently in beta - [nuget.org/packages/Azure.Messaging.EventGrid/4.0.0-beta.2](https://www.nuget.org/packages/Azure.Messaging.EventGrid/4.0.0-beta.2)

If only the v4 was not a beta 😀

**Comment by ahamad-MS** [comment]
Hello @jsgoupil ... Can you please move to use the newest SDK https://www.nuget.org/packages/Azure.Messaging.EventGrid/4.0.0-beta.2 and let us know if you still hitting the issue?

Thanks

**Comment by jsgoupil** [comment]
This is a total new package... With no documentation.
Well, one more in the list... you guys change package once a year, seriously?

There are too many packages to choose correctly, you guys should have a documentation on which package to pick because this is absurd.
* Microsoft.Azure.EventGrid
* Microsoft.Azure.Management.EventGrid
* Microsoft.Azure.WebJobs.Extensions.EventGrid
* Azure.Messaging.EventGrid

Based on this code
https://github.com/Azure/azure-sdk-for-net/blob/b20c241c4eb82a2636cd03afb7eee795abfe4ecc/sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridEvent.cs#L100

It seems that you will handle it.
I'm not updating to your new package though. No doc.
Thanks.

**Comment by ahamad-MS** [comment]
Hello @jsgoupil .. thanks for your feedback and we understand your concern.

FYI, Microsoft.Azure.Management.EventGrid is used for various CRUD management APIs through ARM for Event Grid. These are used for control plane only. Microsoft.Azure.EventGrid and Azure.Messaging.EventGrid are for the data plane operations only. The separation between management/control and data plane is the pattern observed by all Azure services and not specific to Event Grid. Actually, it is a requirement for our team to do so.

As for the new Azure.Messaging.EventGrid.. We are still in the process of authoring/finalizing this new package. We are working on documentation including migration documents. The goal is to ensure the new package addresses all customers feedback and include all new features (as well as fix any existing bugs in the old package).

We appreciate and understand the pain during this period and we are working hard to address this.

thanks
